src/boost_histogram/_internal/deprecated.py

```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+import functools
+import warnings
+
+# Warning: this should not be directly used on properties. It will trigger on
+# tab completion - ALL tab completion that could include this property.
+# ob.<tab> will produce a warning, for example. Instead use a hidden method and
+# a __getattr__ if the property was not settable.
+
+
+class deprecated:
+    def __init__(self, reason, name=""):
+        # type: (str, str) -> None
+        self._reason = reason
+        self._name = name
+
+    def __call__(self, func):
+        @functools.wraps(func)
+        def decorated_func(*args, **kwargs):
+            warnings.warn(
+                "{0} is deprecated: {1}".format(
+                    self._name or func.__name__, self._reason
+                ),
+                category=FutureWarning,
+                stacklevel=2,
+            )
+            return func(*args, **kwargs)
+
+        decorated_func.__doc__ = "DEPRECATED: " + self._reason + "\n" + func.__doc__
+        return decorated_func
```

Comment:
Quick outline of changes from iMinuit:
* I needed a settable name so that I could implement a workaround for properties and tab completion splitting out the deprecation warning even when unrelated.
* I'm using [`functools.wraps`](https://www.geeksforgeeks.org/python-functools-wraps-function/) because it gets lots of properties right, including the signature, at least for `help()` and `inspect.signature`. It's not as good as the [decorator](https://pypi.org/project/decorator/) or [wrapt](https://pypi.org/project/wrapt/) libraries, but I think it's okay for this.
* I'm using FutureWarning, as that's targeted at changes affecting users. DeprecationWarnings is for library authors using your library (who happen to remember to turn them on, they tend to be hidden). [Nice explanation here](https://deprecated.readthedocs.io/en/latest/tutorial.html#deprecation-warning-classes).
* I'm including the original doc string at the end (the user can delete it if they want to).

By the way, [this series of 10 articles](https://github.com/GrahamDumpleton/wrapt/tree/master/blog#readme) on wrapping functions by the wrapt author is fantastic, at least it is if I remember correctly, I think I last read it around the time he wrote it. Not needed for a simple internal decorator like this, but it's why I'd use a proper library for anything public.